README.md
# Deploy Stack - NoSQL Client Server 

This stack creates two Compute Engine Instances, installs MongoDB on one, and 
an API that acts as a client for the database on the other. It connects them
together and sets up firewall rules to expose the API content publically.  

![nosql client server  architecture](architecture.png)

## Install
You can install this application using the `Open in Google Cloud Shell` button 
below. 

<a href="https://ssh.cloud.google.com/cloudshell/editor?cloudshell_git_repo=https%3A%2F%2Fgithub.com%2FGoogleCloudPlatform%2Fdeploystack-nosql-client-server&shellonly=true&cloudshell_image=gcr.io/ds-artifacts-cloudshell/deploystack_custom_image">
    <img alt="Open in Cloud Shell" src="https://gstatic.com/cloudssh/images/open-btn.svg">
</a>

Clicking this link will take you right to the DeployStack app, running in your 
Cloud Shell environment. It will walk you through setting up your architecture.  

## Cleanup 
To remove all billing components from the project
1. Typing `deploystack uninstall`

This is not an official Google product.

CHANGELOG.md
# Changelog

## [1.0.3](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/compare/v1.0.2...v1.0.3) (2023-07-06)


### Bug Fixes

* update ubuntu image to version 20.04 ([fe844da](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/fe844daf5d7e37d7c56f378fd559625e875d6c5c))

## [1.0.2](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/compare/v1.0.1...v1.0.2) (2023-03-20)


### Bug Fixes

* uses new app instead of old version. ([3ca41da](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/3ca41dadd969b5ee309f0e50b29fd4a8d5908e25))

## [1.0.1](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/compare/v1.0.0...v1.0.1) (2023-03-18)


### Bug Fixes

* got a basic http website working with mongo ([8709403](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/87094032588e1a36d78ffdeb5a999b41c5a7c85e))
* initial import of the existing client code ([88b9fac](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/88b9fac47b0ea372cb1e6a09f52d9d50ca7aa7e5))
* pushing last of the app code ([1fce962](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/1fce962dd2d80a5fb103b4aafa3f13bc00fcbf92))
* updated logging for the demo app ([935d713](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/935d713922e23b3f16840b298a9e7c4b5e9472ef))

## 1.0.0 (2023-03-10)


### Features

* adding disposable projects to the mix ([97c70ec](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/97c70ecd1e6c52e84d34f7600b935c17b8e44124))
* adding documentation link. ([f2e72c5](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/f2e72c5995249af0cf4ca2e0b78f7256bcc56008))


### Bug Fixes

* added explicit dependencies on the service being available. ([691c6c2](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/691c6c26457e53c5e7c031d8fdd45b9152efefc5))
* broken testing ([7dd50fd](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/7dd50fd8f2547cbc7fc45bbecaacecd8e03ff796))
* converting to latest preferred config setup ([7c18dae](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/7c18daea07f802c874f7abaa45f527f696b1f91e))
* correcting test path ([8a3cc01](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/8a3cc01ff0cbbaeff6812081b6447604e39033d8))
* Create .github/workflows/main.yml ([979ff26](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/979ff26220cdeb16df2a37049295369de3d33187))
* do not need to clean anymore, because we now use disposable projects. ([591d0d2](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/591d0d223dba22b68e6d7cba0b9843459dce316a))
* fixing broken tests ([20d784f](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/20d784f7ee175803e29ed26842414d99ca5e653f))
* get release please working ([f3f6a62](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/f3f6a626cdf08b1f592729239252e69d61ef6db5))
* getting release please syntax correct. ([60dda8c](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/60dda8c5889ac8d8ddf5bc122eb5101cf675de69))
* getting test project account creation correct ([3cf6500](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/3cf65002e07e0cae8ce0a75b26f812edf168ff42))
* got rid of project number ([ce90c05](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/ce90c051036630da832144b694e851a3b0129cb5))
* improving test procedure ([bf60822](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/bf60822648d999ffe90ea9b190ec4ace6cd3cd2b))
* making sure that app uses private network ([9f691d8](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/9f691d839b0eb746a963ec3cf3da09da454f04bc))
* making sure to clean all of the rules. ([754af03](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/754af0390236c1d4e59b3049ea2a1d62475aad8f))
* moving files around to organize ([689355d](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/689355de06213c41a609724accf4d7bb9c4bab63))
* removing project number as a thing deploystack worries about ([4080e16](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/4080e162218962365fc90a81d1ea88d7565c37d6))
* renaming scripts from deprecated name. ([b53e55e](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/b53e55e61c747ceb27f9be5627c6babc57e1035e))
* set release type ([a36cd0b](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/a36cd0b2876da5f8168cd0912f50443c5be293ca))
* taking advantage of postscript to wait till solution is ready. ([5a81316](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/5a81316cb89d6f135a7e5e715fa4c132144245c3))
* trying to get billing queries to work ([997fbac](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/997fbac2e77ea5c5a12b3f49cd22b7c9866ca244))
* trying to get tests to work properly. ([6245a6b](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/6245a6bd70c6e5d4f2c152b5fd075891ee2c1dfa))
* typo was causing project to fail waiting for client to be active ([03bdf82](https://github.com/GoogleCloudPlatform/deploystack-nosql-client-server/commit/03bdf8298a5ebc11a2129e2940eedbaebe0da701))


code/client/main_test.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

type mockService struct {
	forceErr bool
	jsonErr  bool
}

var errForced = fmt.Errorf("this is a forced error")

func (m mockService) load(context.Context, []trainer) error {
	if m.forceErr {
		return errForced
	}

	return nil
}

func (m mockService) create(context.Context, trainer) error {
	if m.forceErr {
		return errForced
	}
	return nil
}
func (m mockService) delete(context.Context, trainer) error {
	if m.forceErr {
		return errForced
	}
	return nil
}

func (m mockService) list(context.Context) ([]*trainer, error) {
	if m.forceErr {
		return nil, errForced
	}

	trainers := []*trainer{
		{Name: "Ash", Age: 20, City: "Pallet Town"},
		{Name: "Misty", Age: 22, City: "Cerulean City"},
		{Name: "Brock", Age: 35, City: "Pewter City"},
	}

	return trainers, nil
}
func (m mockService) update(context.Context, trainer, trainer) error {
	if m.forceErr {
		return errForced
	}
	return nil
}

func TestHandlers(t *testing.T) {

	tests := map[string]struct {
		in     http.HandlerFunc
		method string
		want   string
		body   string
		status int
	}{
		"healthz": {
			in:     healthHandler,
			method: http.MethodGet,
			status: http.StatusOK,
			want:   `ok`,
		},
		"list": {
			in:     listHandler(mockService{}),
			method: http.MethodGet,
			status: http.StatusOK,
			want:   `[{"name":"Ash","age":20,"city":"Pallet Town"},{"name":"Misty","age":22,"city":"Cerulean City"},{"name":"Brock","age":35,"city":"Pewter City"}]`,
		},
		"create": {
			in:     createHandler(mockService{}),
			method: http.MethodPost,
			status: http.StatusCreated,
			body:   `{"name":"Han","age":33,"city":"Cloud City"}`,
			want:   `{"name":"Han","age":33,"city":"Cloud City"}`,
		},
		"delete": {
			in:     deleteHandler(mockService{}),
			method: http.MethodDelete,
			status: http.StatusNoContent,
			body:   `{"name":"Han","age":33,"city":"Cloud City"}`,
		},
		"update": {
			in:     updateHandler(mockService{}),
			method: http.MethodPut,
			status: http.StatusOK,
			body:   `{"original":{"name":"Han","age":33,"city":"Cloud City"},"replacement":{"name":"Han","age":33,"city":"Cloud City"}}`,
		},
		"listErr": {
			in:     listHandler(mockService{forceErr: true}),
			method: http.MethodGet,
			status: http.StatusInternalServerError,
			want:   `this is a forced error`,
		},
		"createErr": {
			in:     createHandler(mockService{forceErr: true}),
			method: http.MethodPost,
			status: http.StatusInternalServerError,
			body:   `{"name":"Han","age":33,"city":"Cloud City"}`,
			want:   `this is a forced error`,
		},
		"deleteErr": {
			in:     deleteHandler(mockService{forceErr: true}),
			method: http.MethodDelete,
			status: http.StatusInternalServerError,
			body:   `{"name":"Han","age":33,"city":"Cloud City"}`,
			want:   `this is a forced error`,
		},
		"updateErr": {
			in:     updateHandler(mockService{forceErr: true}),
			method: http.MethodPut,
			status: http.StatusInternalServerError,
			body:   `{"original":{"name":"Han","age":33,"city":"Cloud City"},"replacement":{"name":"Han","age":33,"city":"Cloud City"}}`,
			want:   `this is a forced error`,
		},
		"createJSONErr": {
			in:     createHandler(mockService{forceErr: true}),
			method: http.MethodPost,
			status: http.StatusInternalServerError,
			body:   `this aint no json`,
			want:   `invalid character 'h' in literal true (expecting 'r')`,
		},
		"deleteJSONErr": {
			in:     deleteHandler(mockService{forceErr: true}),
			method: http.MethodDelete,
			status: http.StatusInternalServerError,
			body:   `this aint no json`,
			want:   "invalid character 'h' in literal true (expecting 'r')",
		},
		"updateJSONErr": {
			in:     updateHandler(mockService{forceErr: true}),
			method: http.MethodPut,
			status: http.StatusInternalServerError,
			body:   `this aint no json`,
			want:   `invalid character 'h' in literal true (expecting 'r')`,
		},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {

			body := []byte(tc.body)

			req := httptest.NewRequest(tc.method, "/api/v1/trainer", bytes.NewReader(body))
			w := httptest.NewRecorder()
			tc.in(w, req)
			res := w.Result()
			defer res.Body.Close()
			got, err := ioutil.ReadAll(res.Body)

			require.Nil(t, err)
			assert.Equal(t, tc.want, string(got))

		})
	}
}


.deploystack/main_test.go
package main

import (
	"fmt"
	"os"
	"testing"

	"github.com/GoogleCloudPlatform/deploystack"
	"github.com/GoogleCloudPlatform/deploystack/dstester"
)

var (
	ops              = dstester.NewOperationsSet()
	project, _       = deploystack.ProjectID()
	projectNumber, _ = deploystack.ProjectNumber(project)
	basename         = "nosql-client-server"
	debug            = false

	tf = dstester.Terraform{
		Dir: "../terraform",
		Vars: map[string]string{
			"project_id":     project,
			"project_number": projectNumber,
			"region":         "us-central1",
			"zone":           "us-central1-a",
			"basename":       basename,
		},
	}

	resources = dstester.Resources{
		Project: project,
		Items: []dstester.Resource{
			{
				Product: "compute instances",
				Name:    "client",
			},
			{
				Product: "compute instances",
				Name:    "server",
			},
			{
				Product: "compute networks",
				Name:    fmt.Sprintf("%s-network", basename),
			},
			{
				Product: "compute firewall-rules",
				Name:    "deploystack-allow-http",
			},
			{
				Product: "compute firewall-rules",
				Name:    "deploystack-allow-ssh",
			},
			{
				Product: "compute firewall-rules",
				Name:    "deploystack-allow-internal",
			},
		},
	}
)

func init() {
	if os.Getenv("debug") != "" {
		debug = true
	}

	ops.Add("postApply", dstester.Operation{Output: "client_url", Type: "httpPoll"})
}

func TestListCommands(t *testing.T) {
	resources.Init()
	dstester.DebugCommands(t, tf, resources)
}

func TestStack(t *testing.T) {
	dstester.TestStack(t, tf, resources, ops, debug)
}

func TestClean(t *testing.T) {
	if os.Getenv("clean") == "" {
		t.Skip("Clean must be very intentionally called")
	}

	resources.Init()
	dstester.Clean(t, tf, resources)
}


code/client/model_test.go
package main

import (
	"context"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type mockCollection struct {
	forceErr bool
	empty    bool
}

func (m mockCollection) InsertMany(context.Context, []interface{}, ...*options.InsertManyOptions) (*mongo.InsertManyResult, error) {

	result := mongo.InsertManyResult{}

	if m.forceErr {
		return nil, errForced
	}
	return &result, nil

}

func (m mockCollection) InsertOne(context.Context, interface{}, ...*options.InsertOneOptions) (*mongo.InsertOneResult, error) {

	result := mongo.InsertOneResult{}

	if m.forceErr {
		return nil, errForced
	}
	return &result, nil

}

func (m mockCollection) DeleteOne(context.Context, interface{}, ...*options.DeleteOptions) (*mongo.DeleteResult, error) {
	result := mongo.DeleteResult{}

	if m.forceErr {
		return nil, errForced
	}
	return &result, nil

}

func (m mockCollection) ReplaceOne(context.Context, interface{}, interface{}, ...*options.ReplaceOptions) (*mongo.UpdateResult, error) {
	result := mongo.UpdateResult{}

	if m.forceErr {
		return nil, errForced
	}
	return &result, nil

}

func (m mockCollection) Find(context.Context, interface{}, ...*options.FindOptions) (cur *mongo.Cursor, err error) {
	documents := []interface{}{}

	if !m.empty {
		documents = append(documents, trainer{Name: "Ash", Age: 20, City: "Pallet Town"})
		documents = append(documents, trainer{Name: "Misty", Age: 22, City: "Cerulean City"})
		documents = append(documents, trainer{Name: "Brock", Age: 35, City: "Pewter City"})
	}

	result, err := mongo.NewCursorFromDocuments(documents, nil, nil)

	if m.forceErr {
		return nil, errForced
	}
	return result, nil

}

func TestCollection(t *testing.T) {
	tests := map[string]struct {
		collection trainerCollection
		f          func(*trainerService) error
		err        error
	}{
		"load": {
			f: func(svc *trainerService) error {
				err := svc.load(context.Background(), []trainer{
					{Name: "Ash", Age: 20, City: "Pallet Town"},
					{Name: "Misty", Age: 22, City: "Cerulean City"},
					{Name: "Brock", Age: 35, City: "Pewter City"},
				})
				return err
			},
			collection: mockCollection{},
		},
		"loaderr": {
			f: func(svc *trainerService) error {
				err := svc.load(context.Background(), []trainer{
					{Name: "Ash", Age: 20, City: "Pallet Town"},
					{Name: "Misty", Age: 22, City: "Cerulean City"},
					{Name: "Brock", Age: 35, City: "Pewter City"},
				})
				return err
			},
			collection: mockCollection{forceErr: true},
			err:        errForced,
		},
		"loadempty": {
			f: func(svc *trainerService) error {
				err := svc.load(context.Background(), []trainer{
					{Name: "Ash", Age: 20, City: "Pallet Town"},
					{Name: "Misty", Age: 22, City: "Cerulean City"},
					{Name: "Brock", Age: 35, City: "Pewter City"},
				})
				return err
			},
			collection: mockCollection{empty: true},
		},

		"create": {
			f: func(svc *trainerService) error {
				t := trainer{Name: "Ash", Age: 20, City: "Pallet Town"}
				err := svc.create(context.Background(), t)
				return err

			},
			collection: mockCollection{},
		},
		"createerr": {
			f: func(svc *trainerService) error {
				t := trainer{Name: "Ash", Age: 20, City: "Pallet Town"}
				err := svc.create(context.Background(), t)
				return err

			},
			err:        errForced,
			collection: mockCollection{forceErr: true},
		},

		"delete": {
			f: func(svc *trainerService) error {
				t := trainer{Name: "Ash", Age: 20, City: "Pallet Town"}
				err := svc.delete(context.Background(), t)
				return err

			},
			collection: mockCollection{},
		},
		"deleteerr": {
			f: func(svc *trainerService) error {
				t := trainer{Name: "Ash", Age: 20, City: "Pallet Town"}
				err := svc.delete(context.Background(), t)
				return err

			},
			err:        errForced,
			collection: mockCollection{forceErr: true},
		},

		"update": {
			f: func(svc *trainerService) error {
				t := trainer{Name: "Ash", Age: 20, City: "Pallet Town"}
				t2 := trainer{Name: "Ash2", Age: 20, City: "Pallet Town"}
				err := svc.update(context.Background(), t, t2)
				return err

			},
			collection: mockCollection{},
		},
		"updateerr": {
			f: func(svc *trainerService) error {
				t := trainer{Name: "Ash", Age: 20, City: "Pallet Town"}
				t2 := trainer{Name: "Ash2", Age: 20, City: "Pallet Town"}
				err := svc.update(context.Background(), t, t2)
				return err

			},
			err:        errForced,
			collection: mockCollection{forceErr: true},
		},
		"init": {
			f: func(svc *trainerService) error {
				err := initData(context.Background(), svc)
				return err

			},
			collection: mockCollection{},
		},
		"initerr": {
			f: func(svc *trainerService) error {
				err := initData(context.Background(), svc)
				return err

			},
			err:        errForced,
			collection: mockCollection{forceErr: true},
		},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			svc := &trainerService{
				collection: tc.collection,
			}

			err := tc.f(svc)
			if tc.err == nil {
				require.Nil(t, err)
			} else {
				assert.ErrorIs(t, err, tc.err)
			}

		})
	}
}

func TestNewTrainerService(t *testing.T) {
	dl := 1 * time.Second
	ctx, cancel := context.WithTimeout(context.Background(), dl)
	defer cancel()

	_, err := newTrainerService(ctx, "", "")
	want := "error initializing data"

	assert.ErrorContains(t, err, want)

}


code/client/main.go
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package main

import (
	"context"
	"embed"
	"encoding/json"
	"fmt"
	"io/fs"
	"net/http"
	"os"

	"github.com/charmbracelet/log"
	"github.com/gorilla/handlers"
	"github.com/gorilla/mux"
)

//go:embed static
var static embed.FS

const (
	mongoport   = "27017"
	defaultPort = "80"
)

func main() {
	dbHost := os.Getenv("DBHOST")
	port := os.Getenv("PORT")

	if port == "" {
		port = defaultPort
	}

	if dbHost == "" {
		log.Fatal("DBHOST env not set. Need ip/host with mongo on port", mongoport)
	}

	ctx := context.Background()
	svc, err := newTrainerService(ctx, dbHost, mongoport)
	if err != nil {
		log.Fatal("error connecting to mongo: ", err)
	}

	fSys, err := fs.Sub(static, "static")
	if err != nil {
		panic(err)
	}

	router := mux.NewRouter().StrictSlash(true)
	router.HandleFunc("/healthz", healthHandler).Methods(http.MethodGet)
	router.HandleFunc("/api/v1/healthz", healthHandler).Methods(http.MethodGet)
	router.HandleFunc("/api/v1/trainer", listHandler(svc)).Methods(http.MethodGet)
	router.HandleFunc("/api/v1/trainer", createHandler(svc)).Methods(http.MethodPost)
	router.HandleFunc("/api/v1/trainer", deleteHandler(svc)).Methods(http.MethodDelete)
	router.HandleFunc("/api/v1/trainer", updateHandler(svc)).Methods(http.MethodPut)
	router.PathPrefix("/").Handler(logWrap(http.FileServer(http.FS(fSys))))

	headersOk := handlers.AllowedHeaders([]string{"X-Requested-With"})
	originsOk := handlers.AllowedOrigins([]string{"*"})
	methodsOk := handlers.AllowedMethods([]string{"GET", "HEAD", "POST", "PUT", "OPTIONS", "DELETE"})

	log.Fatal(http.ListenAndServe(":"+port, handlers.CORS(originsOk, headersOk, methodsOk)(router)))
}

func respond(w http.ResponseWriter, r *http.Request, status int, content []byte, err error) {
	c := fmt.Sprintf("%s %s %s %s %d", r.Host, r.Method, r.RequestURI, r.Proto, status)
	w.WriteHeader(status)
	if content == nil && err != nil {
		log.With("source", "api").Error(c, "error", err)
		content = []byte(err.Error())
		w.Write(content)
		return
	}

	log.With("source", "api").Info(c)
	w.Write(content)
	return
}
func logWrap(h http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		recorder := &statusRecorder{
			ResponseWriter: w,
			Status:         200,
		}
		h.ServeHTTP(recorder, r)
		c := fmt.Sprintf("%s %s %s %s %d", r.Host, r.Method, r.RequestURI, r.Proto, recorder.Status)
		log.With("source", "static").Info(c)
	})
}

type statusRecorder struct {
	http.ResponseWriter
	Status int
}

func (r *statusRecorder) WriteHeader(status int) {
	r.Status = status
	r.ResponseWriter.WriteHeader(status)
}

func healthHandler(w http.ResponseWriter, r *http.Request) {
	respond(w, r, http.StatusInternalServerError, []byte("ok"), nil)
	return
}

func listHandler(svc trainerCRUDer) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		trainers, err := svc.list(context.Background())
		if err != nil {
			respond(w, r, http.StatusInternalServerError, nil, err)
			return
		}

		j, err := json.Marshal(trainers)
		if err != nil {
			respond(w, r, http.StatusInternalServerError, nil, err)
			return
		}
		respond(w, r, http.StatusOK, j, nil)
		return
	}
}

func createHandler(svc trainerCRUDer) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		t := trainer{}

		if err := json.NewDecoder(r.Body).Decode(&t); err != nil {
			respond(w, r, http.StatusInternalServerError, nil, err)
			return
		}

		if err := svc.create(context.Background(), t); err != nil {
			respond(w, r, http.StatusInternalServerError, nil, err)
			return
		}

		j, err := json.Marshal(t)
		if err != nil {
			respond(w, r, http.StatusInternalServerError, nil, err)
			return
		}
		respond(w, r, http.StatusCreated, j, nil)
		return
	}
}

func deleteHandler(svc trainerCRUDer) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		t := trainer{}

		if err := json.NewDecoder(r.Body).Decode(&t); err != nil {
			respond(w, r, http.StatusInternalServerError, nil, err)
			return
		}

		if err := svc.delete(context.Background(), t); err != nil {
			respond(w, r, http.StatusInternalServerError, nil, err)
			return
		}

		respond(w, r, http.StatusNoContent, nil, nil)
		return
	}
}

func updateHandler(svc trainerCRUDer) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		data := map[string]trainer{}

		if err := json.NewDecoder(r.Body).Decode(&data); err != nil {
			respond(w, r, http.StatusInternalServerError, nil, err)
			return
		}

		if err := svc.update(context.Background(), data["original"], data["replacement"]); err != nil {
			respond(w, r, http.StatusInternalServerError, nil, err)
			return
		}

		respond(w, r, http.StatusOK, nil, nil)
		return
	}
}


code/client/model.go
package main

import (
	"context"
	"fmt"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type trainer struct {
	Name string `json:"name"`
	Age  int    `json:"age"`
	City string `json:"city"`
}

type trainerCRUDer interface {
	load(context.Context, []trainer) error
	list(context.Context) ([]*trainer, error)
	create(context.Context, trainer) error
	delete(context.Context, trainer) error
	update(context.Context, trainer, trainer) error
}

type trainerCollection interface {
	InsertMany(context.Context, []interface{}, ...*options.InsertManyOptions) (*mongo.InsertManyResult, error)
	InsertOne(context.Context, interface{}, ...*options.InsertOneOptions) (*mongo.InsertOneResult, error)
	DeleteOne(context.Context, interface{}, ...*options.DeleteOptions) (*mongo.DeleteResult, error)
	ReplaceOne(context.Context, interface{}, interface{}, ...*options.ReplaceOptions) (*mongo.UpdateResult, error)
	Find(context.Context, interface{}, ...*options.FindOptions) (cur *mongo.Cursor, err error)
}

type trainerService struct {
	collection trainerCollection
}

// newTrainerService spins up a new TrainerManager for interacting with MongoDB.
func newTrainerService(ctx context.Context, host, port string) (*trainerService, error) {
	uri := fmt.Sprintf("mongodb://%s:%s", host, port)
	clientOptions := options.Client().ApplyURI(uri)

	client, err := mongo.Connect(ctx, clientOptions)
	if err != nil {
		return nil, fmt.Errorf("error connecting to mongo: %w", err)
	}

	collection := client.Database("test").Collection("trainers")

	svc := &trainerService{
		collection: collection,
	}

	if err := initData(ctx, svc); err != nil {
		return nil, fmt.Errorf("error initializing data: %w", err)
	}

	return svc, nil
}

func initData(ctx context.Context, svc trainerCRUDer) error {
	trainers := []trainer{
		{Name: "Ash", Age: 20, City: "Pallet Town"},
		{Name: "Misty", Age: 22, City: "Cerulean City"},
		{Name: "Brock", Age: 35, City: "Pewter City"},
	}

	return svc.load(ctx, trainers)
}

// load pushes a collection of trainers into a mongoDB instance
func (svc *trainerService) load(ctx context.Context, trainers []trainer) error {
	t := make([]interface{}, len(trainers))
	for i, tdata := range trainers {
		t[i] = tdata
	}

	list, err := svc.list(ctx)
	if err != nil {
		return fmt.Errorf("error checking before loading to mongo: %w", err)
	}

	if len(list) > 0 {
		return nil
	}

	if _, err := svc.collection.InsertMany(ctx, t); err != nil {
		return fmt.Errorf("error inserting records to mongo: %w", err)
	}

	return nil
}

func (svc *trainerService) create(ctx context.Context, trainer trainer) error {

	if _, err := svc.collection.InsertOne(ctx, trainer); err != nil {
		return fmt.Errorf("error inserting record to mongo: %w", err)
	}

	return nil
}

func (svc *trainerService) delete(ctx context.Context, trainer trainer) error {
	if _, err := svc.collection.DeleteOne(ctx, trainer); err != nil {
		return fmt.Errorf("error inserting record to mongo: %w", err)
	}
	return nil
}

func (svc *trainerService) update(ctx context.Context, original, replacement trainer) error {
	if _, err := svc.collection.ReplaceOne(ctx, original, replacement); err != nil {
		return fmt.Errorf("error replacing record in mongo: %w", err)
	}
	return nil
}

// list retrieves the total collection of trainers from a mongoDB instance
func (svc *trainerService) list(ctx context.Context) ([]*trainer, error) {
	var results []*trainer

	cur, err := svc.collection.Find(ctx, bson.D{{}}, options.Find())
	if err != nil {
		return nil, err
	}
	defer cur.Close(ctx)

	for cur.Next(ctx) {
		var elem trainer
		if err := cur.Decode(&elem); err != nil {
			return nil, err
		}
		results = append(results, &elem)
	}

	return results, cur.Err()
}
